src/child-process.coffee

```diff
@@ -0,0 +1,182 @@
+    # the very same character to stdin of the process, it's not
+    # recognized (the rl.on('SIGINT') event won't get triggered)
+    # for some reason.
+    #
+    # The only thing Dredd is left with is a convention. So when Dredd
+    # wants to gracefully signal to the child process it should terminate,
+    # it sends the '\u0003' to stdin of the child. It's up to the child
+    # to implement reading from stdin in such way it works both for
+    # programmatic and manual Ctrl+C.
+    childProcess.stdin.write(String.fromCharCode(ASCII_CTRL_C))
+  else
+    childProcess.kill('SIGTERM')
+  process.nextTick(callback)
+
+
+# Gracefully terminates a child process
+#
+# Sends a signal to the process as a heads up it should terminate.
+# Then checks multiple times whether the process terminated. Retries
+# sending the signal. In case it's not able to terminate the process
+# within given timeout, it returns an error.
+#
+# If provided with the 'force' option, instead of returning an error,
+# it kills the process unconditionally.
+#
+# Available options:
+# - timeout (number) - Time period in ms for which the termination
+#                      attempts will be done
+# - retryDelay (number) - Delay in ms between termination attempts
+# - force (boolean) - Kills the process forcefully after the timeout
+terminate = (childProcess, options = {}, callback) ->
+  [callback, options] = [options, {}] if typeof options is 'function'
+  force = options.force or false
+
+  # If the timeout is zero or less then the delay for waiting between
+  # retries, there will be just one termination attempt
+  timeout = if options.timeout? then options.timeout else TERM_DEFAULT_TIMEOUT_MS
+  retryDelay = if options.retryDelay? then options.retryDelay else TERM_DEFAULT_RETRY_MS
+
+  terminated = false
+  onClose = ->
+    terminated = true
+    childProcess.removeListener('close', onClose)
+  childProcess.on('close', onClose)
+
+  start = Date.now()
+  t = undefined
+
+  # A function representing one check, whether the process already
+  # ended or not. It is repeatedly called until the timeout has passed.
+  check = ->
+    if terminated
+      # Successfully terminated
+      clearTimeout(t)
+      callback()
+    else
+      if (Date.now() - start) < timeout
+        # Still not terminated, try again
+        signalTerm(childProcess, (err) ->
+          return callback(err) if err
+          t = setTimeout(check, retryDelay)
+        )
+      else
+        # Still not terminated and the timeout has passed, either
+        # kill the process (force) or provide an error
+        clearTimeout(t)
+        if force
+          signalKill(childProcess, callback)
+        else
+          callback(new Error("Unable to gracefully terminate process #{childProcess.pid}"))
+
+  # Fire the first termination attempt and check the result
+  signalTerm(childProcess, (err) ->
+    return callback(err) if err
+    t = setTimeout(check, TERM_FIRST_CHECK_TIMEOUT_MS)
```

Comment:
Basically, yes. It's there so the `check` could be simpler, because it can rely that the first iteration is no different from any other iteration.